src/coreclr/jit/gentree.cpp

```diff
@@ -22865,7 +22865,7 @@ bool GenTreeLclFld::IsOffsetMisaligned() const
 
 bool GenTree::IsInvariant() const
 {
-    return OperIsConst() || Compiler::impIsAddressInLocal(this);
+    return OperIsConst() || IsLocalAddrExpr();
```

Comment:
It's notable that combined with #68469, this will bind `IsLocalAddrExpr` "the simple version" to `LocalAddressVisitor` like `DefinesLocalAddr` and `IsLocalAddrExpr` "the not simple version" are bound right now (which may also not be a new thing, it may have been like this already due to struct promotion requirements).

BTW, this is also the reason we _must_ recognize `ADD`s in all these functions, they're produced for structs with fields at offsets larger than `GenTreeLclFld` (and the emitter) can represent.

Reply:
I think even without #68469 there are/were non-trivial invariants between the two. While before this PR we 'supported' call morphing introducing a local store for the address, there was already an implicit assumption that no extra uses of this temp were ever created. So in a way we are just trading one assumption for another.

Reply:
This is unsatisfying, but I am hard pressed to think about how to improve the situation.
Probably it would be ideal if we could unify `DefinesLocalAddr`, the two variants of `IsLocalAddrExpr` and what `LocalAddressVisitor` is doing somehow, but that also seems difficult.

Reply:
Hehe, this is what #11057 is about :).

The original "north star" design was that we'd only have `STORE_LCL_VAR|FLD` for tracked locals, this return buffer optimization makes that "hard" unfortunately, but even with it (assuming I won't find a way to represent it as a store after all), we'd have:
1) `DefinesLocalAddr` becomes equivalent to `OperIsLocalAddr`.
2) All its variations (`IsLocalAddrExpr` "simple" and "not simple", as well as `fgIsIndirOfAddrOfLocal`) are deleted.
3) `LocalAddressVisitor` transforms all `ASG`s into `STORE_LCL_VAR|FLD`, and for the return buffer optimization, `LCL_VAR|FLD_ADDR`.
4) All indirect uses become `LCL_VAR|FLD` (this is #51569)

As a point-in-time improvement, the "not simple" `IsLocalAddrExpr` will be deleted once/if physical VN changes are in.